python/ray/experimental/sgd/pytorch/pytorch_trainer.py

```diff
@@ -0,0 +1,115 @@
+
+    Launches a set of actors which connect via distributed PyTorch and
+    coordinate gradient updates to train the provided model.
+    """
+
+    def __init__(
+            self,
+            model_creator,
+            data_creator,
+            optimizer_creator=None,  # SGD as default TODO: handle this
+            config=None,
+            num_replicas=1,
+            resources_per_replica=None,
+            batch_size=16,
+            backend="gloo"):
+        """Sets up the PyTorch trainer.
+
+        Args:
+            model_creator (dict -> torch.nn.Module): creates the model
+                using the config.
+            data_creator (dict -> Dataset, Dataset): creates the training
+                and validation data sets using the config.
+            optimizer_creator (model, dict -> loss, optimizer): creates the
+                loss and optimizer using the config.
+            config (dict): configuration passed to 'model_creator',
+                'data_creator', and 'optimizer_creator'.
+            batch_size (int): batch size used for SGD.
+            backend (string): backend used for distributed SGD. "gloo" or
+                "nccl".
+        """
+        # TODO: add support for mixed precision
+        # TODO: add support for callbacks
+        if sys.platform == "darwin":
+            raise Exception(
+                ("Distributed PyTorch is not supported on macOS. For more "
+                 "information, see "
+                 "https://github.com/pytorch/examples/issues/467."))
+
+        self.model_creator = model_creator
+        self.config = {} if config is None else config
+        self.optimizer_timer = utils.TimerStat(window_size=1)
+
+        if resources_per_replica is None:
+            resources_per_replica = utils.Resources(
+                num_cpus=0, num_gpus=1, resources={})
+
+        Runner = ray.remote(
+            num_cpus=resources_per_replica.num_cpus,
+            num_gpus=resources_per_replica.num_gpus,
+            resources=resources_per_replica.resources)(PyTorchRunner)
+
+        self.workers = [
+            Runner.remote(model_creator, data_creator, optimizer_creator,
+                          self.config, batch_size, backend)
+            for _ in range(num_replicas)
+        ]
+
+        ip = ray.get(self.workers[0].get_node_ip.remote())
+        port = utils.find_free_port()
+        address = "tcp://{ip}:{port}".format(ip=ip, port=port)
+        for i, worker in enumerate(self.workers):
+            worker.setup.remote(address, i, len(self.workers))
+
+    def train(self):
+        """Runs a training epoch"""
+        with self.optimizer_timer:
+            worker_stats = ray.get([w.step.remote() for w in self.workers])
+        return worker_stats[0]  # TODO: merge worker stats
+
+    def validate(self):
+        """Evaluates the model on the validation data set"""
+        return ray.get(self.workers[0].validate.remote())
```

Comment:
Are you sure this is doing the right thing? You set distributed sampler for the validation dataset, but only call validate on one method.